### PR TITLE
Parser

### PR DESCRIPTION
**Pull request by WeiqunZhang**
Add amrex::Parser that can be used at runtime to evaluate math expressions
given by strings.  It works on both host and device.  This is moved from
WarpX with additional performance optimization of eliminating recursion.

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [ ] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] are described in the proposed changes to the AMReX documentation, if appropriate

### CODE REVIEW

**Comment by drummerdoc** [comment]
Nice!

File: Src/Base/Parser/README [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+   WarpX, because the files have already been generated.
+
+** wp_parser.tab.c & wp_parser.tab.h
+
+   They are generated by bison.
+
+** wp_parser_y.c & wp_parser_y.h
+
+   These contain C codes that are used to evaluate a mathematical
+   expression given in string format.
```
Comment (jmsexton03) [inline Src/Base/Parser/README]: WarpX -> AMReX ?

Comment (WeiqunZhang) [reply Src/Base/Parser/README]: Thanks.  Forgot to update the README.

**Comment by WeiqunZhang** [comment]
Added support for local variables so that one can do things like `r2=x**2+y**2; r=sqrt(r2); r2*log(r)`.

**Review by atmyers — APPROVED** [review]

**Pull request merged**